src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java

```diff
@@ -2095,209 +2103,230 @@ public void handle(Either<String, JsonArray> responseQuerry) {
+                    Double moy = utilsService.calculMoyenne(noteDevoirList, stats, Field.DIVISEUR_NOTE, annual).getDouble(Field.MOYENNE);
+                    setMapIdEleveMatMoy(mapIdEleveIdMatMoy, moy, idEleve, idMat);
+                }
+            }
+        }
+    }
+    private void calculateAveragesSubjectsWithSubsubjects (Map<String, Map<String, Double>> mapIdEleveIdMatMoy,
+                            HashMap<String, HashMap<String, HashMap<Long, List<NoteDevoir>>>> mapIdEleveIdMatIdSousMatListNotes,
+                            Boolean stats, Boolean annual) {
+        for (Map.Entry<String, HashMap<String, HashMap<Long, List<NoteDevoir>>>> eleveMapEntry : mapIdEleveIdMatIdSousMatListNotes.entrySet()) {
+            for (Map.Entry<String, HashMap<Long, List<NoteDevoir>>> matMapEntry : eleveMapEntry.getValue().entrySet()) {
+                double total = 0;
+                double totalCoeff = 0;
+                String idEleve = eleveMapEntry.getKey();
+                String idMat = matMapEntry.getKey();
+                for (Map.Entry<Long, List<NoteDevoir>> sousMatMapEntry : matMapEntry.getValue().entrySet()) {
+                    Long idSousMat = sousMatMapEntry.getKey();
+                    Service serv = sousMatMapEntry.getValue().get(0).getService();
+                    double coeff = 1.d;
 
-                        elevesResult.add(eleveJsonO);
+                    if (serv != null && serv.getSubtopics() != null && serv.getSubtopics().size() > 0) {
+                        SubTopic subTopic = serv.getSubtopics().stream()
+                                .filter(el ->
+                                        el.getId().equals(idSousMat)
+                                ).findFirst().orElse(null);
+                        if (subTopic != null)
+                            coeff = subTopic.getCoefficient();
                     }
 
-                    JsonObject resultMoysElevesByMat = new JsonObject();
-                    setJOResultWithList(resultMoysElevesByMat, listMoyGeneraleEleve);
-                    elevesResult = Utils.sortElevesByDisplayName(elevesResult);
-                    resultMoysElevesByMat.put("eleves", elevesResult);
-                    resultMoysElevesByMat.put("nbEleves", elevesResult.size());
-                    handler.handle(new Either.Right<>(resultMoysElevesByMat));
+                    Double moyenSousMat = utilsService.calculMoyenne(sousMatMapEntry.getValue(), stats, Field.DIVISEUR_NOTE, annual)
+                            .getDouble(Field.MOYENNE);
+
+                    total += coeff * moyenSousMat;
+                    totalCoeff += coeff;
+                }
+                if (totalCoeff != 0) {
+                    Double moy = Math.round((total / totalCoeff) * Field.ROUNDER) / Field.ROUNDER;
+                    setMapIdEleveMatMoy(mapIdEleveIdMatMoy, moy, idEleve, idMat);
                 }
             }
-        };
+        }
```

Comment:
Voici une suggestion pour utiliser `flatMap ` au lieu du nested for (for dans un for...) : 

```
    private void calculateAveragesSubjectsWithSubsubjects (Map<String, Map<String, Double>> mapIdEleveIdMatMoy,
                            HashMap<String, HashMap<String, HashMap<Long, List<NoteDevoir>>>> mapIdEleveIdMatIdSousMatListNotes,
                            Boolean stats, Boolean annual) {
        mapIdEleveIdMatIdSousMatListNotes.entrySet().stream()
                .flatMap(eleveMapEntry -> eleveMapEntry.getValue().entrySet().stream())
                .forEach(matMapEntry -> {
                    double total = 0;
                    double totalCoeff = 0;
                    String idEleve = matMapEntry.getKey();
                    String idMat = matMapEntry.getKey();
                    for (Map.Entry<Long, List<NoteDevoir>> sousMatMapEntry : matMapEntry.getValue().entrySet()) {
                        Long idSousMat = sousMatMapEntry.getKey();
                        Service serv = sousMatMapEntry.getValue().get(0).getService();
                        double coeff = 1.d;

                        if (serv != null && serv.getSubtopics() != null && !serv.getSubtopics().isEmpty()) {
                            SubTopic subTopic = serv.getSubtopics().stream()
                                    .filter(el ->
                                            el.getId().equals(idSousMat)
                                    ).findFirst().orElse(null);
                            if (subTopic != null)
                                coeff = subTopic.getCoefficient();
                        }

                        Double moyenSousMat = utilsService.calculMoyenne(sousMatMapEntry.getValue(), stats, Field.DIVISEUR_NOTE, annual)
                                .getDouble(Field.MOYENNE);

                        total += coeff * moyenSousMat;
                        totalCoeff += coeff;
                    }
                    if (totalCoeff != 0) {
                        Double moy = Math.round((total / totalCoeff) * Field.ROUNDER) / Field.ROUNDER;
                        setMapIdEleveMatMoy(mapIdEleveIdMatMoy, moy, idEleve, idMat);
                    }
                });
    }
```

On en profite pour remplacer ton instruction `serv.getSubtopics().size() > 0` par ` !serv.getSubtopics().isEmpty()` comme il faut utiliser les méthodes de la collection Array de java